libraries/plugins/transaction_status/transaction_status_plugin.cpp

```diff
@@ -0,0 +1,110 @@
+#include <steem/plugins/transaction_status/transaction_status_plugin.hpp>
+#include <steem/plugins/transaction_status/transaction_status_objects.hpp>
+#include <steem/chain/database.hpp>
+#include <steem/chain/index.hpp>
+
+#include <fc/io/json.hpp>
+
+#define TRANSACTION_STATUS_BLOCK_DEPTH_KEY     "transaction-status-block-depth"
+#define TRANSACTION_STATUS_DEFAULT_BLOCK_DEPTH  64000
+
+namespace steem { namespace plugins { namespace transaction_status {
+
+namespace detail {
+
+class transaction_status_impl
+{
+public:
+   transaction_status_impl() : _db( appbase::app().get_plugin< steem::plugins::chain::chain_plugin >().db() ) {}
+   virtual ~transaction_status_impl() {}
+
+   void on_post_apply_transaction( const transaction_notification& note );
+   void on_post_apply_block( const block_notification& note );
+
+   chain::database&              _db;
+   uint32_t                      block_depth;
+   boost::signals2::connection   post_apply_transaction_connection;
+   boost::signals2::connection   post_apply_block_connection;
+};
+
+void transaction_status_impl::on_post_apply_transaction( const transaction_notification& note )
+{
+   _db.create< transaction_status_object >( [&]( transaction_status_object& obj )
+   {
+      obj.transaction_id = note.transaction_id;
+   } );
+}
+
+void transaction_status_impl::on_post_apply_block( const block_notification& note )
+{
+   // Update all status objects with the transaction current block number
+   for ( const auto& e : note.block.transactions )
+   {
+      const auto& tx_status_obj = _db.get< transaction_status_object, by_trx_id >( e.id() );
+      _db.modify( tx_status_obj, [&] ( transaction_status_object& obj )
+      {
+         obj.block_num = note.block_num;
+      } );
+   }
+
+   // Remove elements from the index that are deemed too old for tracking
+   if ( note.block_num > block_depth )
+   {
+      uint32_t earliest_tracked_block = note.block_num - block_depth;
+      const auto& idx = _db.get_index< transaction_status_index >().indices().get< by_block_num >();
```

Comment:
Removing in a loop like this can be tricky. We haven't had any problems with concurrent modification, but the boost multi-index container documentation doesn't explicitly state support for concurrent modification. Because of this, I would restructure the loop as follows:

``` c++
auto itr = idx.begin();
while( itr != idx.end() && itr->block_num < earliest_tracked_block )
{
   _db.remove( *itr );
   itr = idx.begin();
}
```

This would also require changing the behavior of the block num field slightly to support a non-zero sentinel value such as `#define TX_IN_MEMPOOL ~0`